### ISSUE TITLE
luajit requires allocated addresses to be in the lower 4G range

### ISSUE THREAD

**Issue by tgrabiec** [cloudius-systems/osv#411]
The `wrk` tool (https://github.com/wg/wrk) uses `luajit` (it ships with it). luajit, even when built in 64bit mode, requires that allocation area resides in lower 32-bits of address space, however on OSv addresses start at `0x200000000000`. See `lua_newstate()`:

``` c
if (GG == NULL || !checkptr32(GG)) return NULL;
```

This is making `wrk` page faults very early on zero page access.

When I apply this hack to OSv then wrk starts to work:

```
diff --git a/core/mmu.cc b/core/mmu.cc
index 3b631a9..d2495c9 100644
--- a/core/mmu.cc
+++ b/core/mmu.cc
@@ -983,7 +983,7 @@ uintptr_t allocate(vma *v, uintptr_t st
     if (search) {
         // search for unallocated hole around start
         if (!start) {
-            start = 0x200000000000ul;
+            start = 0x2000000ul;
         }
         start = find_hole(start, size);
     } else {
```

On Linux it seems to start around `0x4000000`.

**Comment by justinc1** [comment]
I believe this issue is back. The wrk/src/script.c:
```
ua_State *L = luaL_newstate();
luaL_openlibs(L);
```
sets L to NULL, and luaL_openlibs than crashes OSv.
Tracking trough function calls, ls_state.c:
```
#if LJ_64
lua_State *lj_state_newstate(lua_Alloc f, void *ud)
#else
LUA_API lua_State *lua_newstate(lua_Alloc f, void *ud)
#endif
{
  GG_State *GG = (GG_State *)f(ud, NULL, 0, sizeof(GG_State));
  lua_State *L = &GG->L;
  global_State *g = &GG->g;
  if (GG == NULL || !checkptr32(GG)) {
    fprintf(stderr, "DBG %s:%d %s, GG=%p, checkptr32 failed? \n", __FILE__, __LINE__, __FUNCTION__, GG);
    return NULL;
  }
```
Prints:
```
DBG lj_state.c:191 lj_state_newstate, GG=0x200000200378, checkptr32 failed?
```
The GG in valid, and checkptr32 failed.

**Comment by justinc1** [comment]
...
Oh, i see now - the patch mentioned above was not included into OSv master, it is supposed to be  applied by the user. Hm, a nice error and exit(1) patch for wrk would help (because gdb trace after crash is useless, as stack is already corrupted).

**Comment by nyh** [comment]
@avikivity do you remember if there's a reason why you committed this patch to your private branch and not to the main OSv branch? This issue was closed prematurely, because the patch was never committed to the main branch.

Avi, your approach looks reasonable to me, although personally I wouldn't have changed so much of the internals - I think that it's enough that mmap() in libc/mman.c checks if MAP_32BIT (and not MAP_FIXED, and perhaps !addr) and if it is, just sets addr to 0x20000 (or whatever) and just go on with the regular code, with no need to add new flags, new parameters to allocate(), changing mmu::map_*(), etc. I think it would have been simpler.

**Comment by nyh** [comment]
@justinc1 @avikivity I think the following patch is much simpler and shorter than Avi's, and makes more sense to me (don't add the ugly Linux hack to the internal API). @justinc1 can you please confirm that it works too?
```patch
diff --git a/libc/mman.cc b/libc/mman.cc
index 0441265..73a3f41 100644
--- a/libc/mman.cc
+++ b/libc/mman.cc
@@ -130,6 +130,13 @@ void *mmap(void *addr, size_t length, int prot, int flags,
     auto mmap_flags = libc_flags_to_mmap(flags);
     auto mmap_perm  = libc_prot_to_perm(prot);
 
+    if ((flags & MAP_32BIT) && !(flags & MAP_FIXED) && !addr) {
+        // If addr is not specified, OSv by default starts mappings at address
+        // 0x200000000000ul (see mmu::allocate()).  MAP_32BIT asks for a lower
+        // default. If MAP_FIXED or addr were specified, the default does not
+        // matter anyway.
+        addr = (void*)0x2000000ul;
+    }
     if (flags & MAP_ANONYMOUS) {
         // We have already determined (see below) the region where the heap must be located. Now the JVM will request
         // fixed mappings inside that region
```

**Comment by avikivity** [comment]
@nyh I don't remember. Your patch is simpler and should work.

**Comment by nyh** [comment]
Fixed by commit 0c3d1989d8bb1978c9d236b076de0b2b97058906
